include/boundary_op.hxx

```diff
@@ -73,13 +66,49 @@ public:
 
   BoundaryRegion *bndry;
   bool apply_to_ddt; // True if this boundary condition should be applied on the time derivatives, false if it should be applied to the field values
+
+protected:
+  std::shared_ptr<FieldGenerator> gen; // Generator
+
+  // Apply boundary condition at a point
+  virtual void applyAtPoint(Field2D &UNUSED(f), BoutReal UNUSED(val), int
+      UNUSED(x), int UNUSED(bx), int UNUSED(y), int UNUSED(by), int UNUSED(z),
+      Coordinates* UNUSED(metric)) {
+    ASSERT1(false);
```

Comment:
This might lead to a confused user as the error message won't say what is wrong just that the assert failed so a throw might be better.

I think if you want it to be an error to not override these routines then you can make these be pure virtual, e.g.

~~~~
virtual void applyAtPoint(Field2D &f, BoutReal val, int x, int bx, int y, int by, int z,
      Coordinates* metric) = 0;
~~~~

and it's then a compile time error.

I notice you pass in the metric here, but you're also passing a field -- could we just get the metric from the field?